### PR TITLE
"$this->"" needs to become "static::" instead of "self::"

### PR DESCRIPTION
**Pull request by derrickschoen**
If a class could be inherited, "$this->"" needs to become "static::" instead of "self::"

### CODE REVIEW

**Comment by TomasVotruba** [comment]
I need more information about this.

**Comment by derrickschoen** [comment]
> I need more information about this.

```
<?php

class SomeClass
{
    public function run()
    {
        $this->eat();
    }
    public static function eat()
    {
        echo "parent";
    }
}

class SomeChild extends SomeClass
{

    public static function eat()
    {
        echo "child";
    }
}

$c = new SomeChild();
$c->run();
```

This echos "child". 

If I replace  ```$this->eat();``` with ```self::eat();```: it echos "parent" which is different than the original output.


However, if I replace  ```$this->eat();``` with ```static::eat();```: it matches the output of the original code: "child"

**Comment by TomasVotruba** [comment]
I see. We should have this covered with a test case.

Could you update the test with Runnable interface? 
https://github.com/rectorphp/rector/blob/f2eb777f1da530db0c5f9201cf6d6248301505bb/rules/code-quality/tests/Rector/FuncCall/ArrayKeysAndInArrayToArrayKeyExistsRector/Fixture/fixture.php.inc#L7

The `run()` method should return value. It will be compared with value after the change.

File: rules/php70/tests/Rector/MethodCall/ThisCallOnStaticMethodToStaticCallRector/Fixture/fixture.php.inc [hunk 1]
```diff
@@ -20,15 +40,35 @@ class SomeClass
...
+}
 
-    public static function eat(string $thing)
+class SomeRunnable implements RunnableInterface
+{
+    public function run()
     {
+        $expectedString = "child";
+        $c = new SomeChild();
+        return $c->speak() === $expectedString;
```
Comment (TomasVotruba) [inline rules/php70/tests/Rector/MethodCall/ThisCallOnStaticMethodToStaticCallRector/Fixture/fixture.php.inc:71]: This should return the value:

```suggestion
        return $c->speak();
```

File: rules/php70/tests/Rector/MethodCall/ThisCallOnStaticMethodToStaticCallRector/Fixture/fixture.php.inc [hunk 2]
```diff
@@ -2,15 +2,35 @@
...
```
Comment (TomasVotruba) [inline rules/php70/tests/Rector/MethodCall/ThisCallOnStaticMethodToStaticCallRector/Fixture/fixture.php.inc:33]: So should this:

```suggestion
        return $c->speak();
```

**Comment by derrickschoen** [comment]
I'm not really sure what I'm doing here. 
Mostly just tried to copy what I saw in Rector\CodeQuality\Tests\Rector\FuncCall\ArrayKeysAndInArrayToArrayKeyExistsRector\Fixture

Wondering why I am getting the error:
```
Cannot declare class Rector\Php70\Tests\Rector\MethodCall\ThisCallOnStaticMethodToStaticCallRector\Fixture\ClassName_t3377givb1400qffpo7a, because the name is already in use in /tmp/rector_temp_tests/runnable_

```

**Comment by TomasVotruba** [comment]
That will be something related to autoloading. I'll handle it tomorrow, don't worry

Thanks for great work :+1:

**Comment by TomasVotruba** [comment]
It took a bit more improvements on runnable testing side, but all works now :)

**Pull request merged**